### PR TITLE
[rush] Preserve pnpm credential environment variables on POSIX

### PR DESCRIPTION
**Pull request by mikedelgaudio**
## Summary

Fix missing pnpm registry authentication on POSIX when `provideNpmrcCredentialsViaEnvironment` is enabled.

Follow-up to #5995: Rush creates URL-scoped `npm_config_*` credential variables, but passing them through `/bin/sh` can discard them before pnpm starts. On Linux with dash, a clean install using pnpm 10.34.5 fails with `ERR_PNPM_FETCH_401` and no Authorization header.

## Details

- Add an internal `useShell` execution option, defaulting to existing shell behavior.
- Opt package-manager launches out of the POSIX shell only when the existing credential-environment experiment/version gate applies. Pass the executable and original argument array directly to the child process.
- Cover workspace installs, legacy installs, autoinstaller installation/update, and `rush-pnpm`.
- Preserve Windows shell handling for `.cmd` shims, default behavior for other callers, output handling, retries, and exit-code capture.
- Keep credentials in the child environment; do not introduce credential files or shell overrides.
- Include the Rush patch release note.

## How it was tested

Revalidated the unchanged patch on macOS with Node 22.21.1:

- `rush --quiet test --only @microsoft/rush-lib --include-phase-deps --parallelism 4`: 931 tests passed, zero failures.
- Six new real-subprocess regressions cover generated credential preservation, the install retry execution path, literal POSIX arguments, exit-code capture, failure rejection, and unchanged default shell execution.
- Whole-suite Rush-library coverage: 63.89% lines and 77.87% branches (not changed-lines coverage).
- `rush change --verify --target-branch origin/main`, Prettier checking all changed files, and `git diff --check origin/main...HEAD` passed.

Prior Linux validation of this exact patch, recorded in the implementation handoff:

- Linux/dash with Node 22.16.0: the same Rush-library command passed all 931 tests.
- Manual clean authenticated-registry comparison with pnpm 10.34.5, isolated HOME/package stores, an empty user npm config, and synthetic credentials: released Rush 5.179.0 failed with 401; patched Rush installed successfully without writing the resolved token to generated config.
- Patched `rush-pnpm outdated` and autoinstaller update authenticated successfully.

The manual Linux registry reproduction was not rerun on macOS. Windows was not executed locally; its existing shell path is unchanged and the POSIX-only argument test is skipped on Windows. Full legacy-install and cold autoinstaller-install integrations were not separately exercised.

### CODE REVIEW

File: common/changes/@microsoft/rush/fix-pnpm-posix-credential-environment_2026-09-18.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Fix pnpm registry credentials being dropped by POSIX shells when provideNpmrcCredentialsViaEnvironment is enabled.",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/fix-pnpm-posix-credential-environment_2026-09-18.json:5]:
```suggestion
      "comment": "Fix pnpm registry credentials being dropped by POSIX shells when `provideNpmrcCredentialsViaEnvironment` is enabled.",
```

File: libraries/rush-lib/src/utilities/test/Utilities.test.ts [hunk 2]
```diff
@@ -23,6 +28,152 @@ function withComSpec<T>(value: string | undefined, callback: () => T): T {
...
 describe(Utilities.name, () => {
+  describe('package manager credential environment', () => {
+    const credentialKey: string = 'npm_config_//registry.example.test/npm/:_authToken';
+    const credentialValue: string = 'non-secret-test-token';
+    let directory: string;
+    let scriptPath: string;
+    let environment: NodeJS.ProcessEnv;
+
+    beforeAll(async () => {
+      directory = await fs.promises.mkdtemp(`${os.tmpdir()}/rush credentials `);
```
Comment (iclanton) [inline libraries/rush-lib/src/utilities/test/Utilities.test.ts:39]: Don't use `os.tempdir`. We want to keep all writes in deterministic places.

File: libraries/rush-lib/src/utilities/test/Utilities.test.ts [hunk 3]
```diff
@@ -1,7 +1,12 @@
 // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 // See LICENSE in the project root for license information.
 
+import * as fs from 'node:fs';
```
Comment (iclanton) [inline libraries/rush-lib/src/utilities/test/Utilities.test.ts:4]: Use `FileSystem` from `node-core-library`.

File: libraries/rush-lib/src/utilities/test/Utilities.test.ts [hunk 4]
```diff
@@ -23,6 +28,152 @@ function withComSpec<T>(value: string | undefined, callback: () => T): T {
...
+      directory = await fs.promises.mkdtemp(`${os.tmpdir()}/rush credentials `);
+      scriptPath = `${directory}/check credentials.cjs`;
+      const sourceFolder: string = `${directory}/source`;
+      const targetFolder: string = `${directory}/target`;
+      await fs.promises.mkdir(sourceFolder);
+      await fs.promises.mkdir(targetFolder);
+      await fs.promises.writeFile(
+        `${sourceFolder}/.npmrc`,
+        '//registry.example.test/npm/:_authToken=${RUSH_TEST_TOKEN}\n'
+      );
```
Comment (iclanton) [inline libraries/rush-lib/src/utilities/test/Utilities.test.ts:48]: `FileSystem.writeFile` has an option to ensure the folder exists.

File: libraries/rush-lib/src/utilities/test/Utilities.test.ts [hunk 5]
```diff
@@ -23,6 +28,152 @@ function withComSpec<T>(value: string | undefined, callback: () => T): T {
...
+          environment,
+          keepEnvironment: true,
+          useShell: false,
+          suppressOutput: true
+        },
+        1
+      );
+    });
+
+    (IS_WINDOWS ? it.skip : it)(
```
Comment (iclanton) [inline libraries/rush-lib/src/utilities/test/Utilities.test.ts:110]: Ensure the cache key for this project's tests takes the OS into account.

File: libraries/rush-lib/src/utilities/test/Utilities.test.ts (hunk 4 above)
Comment (mikedelgaudio) [reply libraries/rush-lib/src/utilities/test/Utilities.test.ts]: Added `ensureFolderExists: true` to both fixture writes in 8866cb5ec3 and removed the separate directory-creation calls. Confirmed `syncNpmrc` already creates its target folder. The updated fixture passes in the full 931-test Rush-library run.

File: libraries/rush-lib/src/utilities/test/Utilities.test.ts (hunk 3 above)
Comment (mikedelgaudio) [reply libraries/rush-lib/src/utilities/test/Utilities.test.ts]: Replaced the direct filesystem calls with `FileSystem.writeFileAsync`, `readFileAsync`, and `deleteFolderAsync` from `@rushstack/node-core-library` in 8866cb5ec3. Removed the `node:fs` and `node:os` imports. The real subprocess tests and generated-config credential assertion remain intact; all 931 Rush-library tests pass.

File: libraries/rush-lib/src/utilities/test/Utilities.test.ts (hunk 5 above)
Comment (mikedelgaudio) [reply libraries/rush-lib/src/utilities/test/Utilities.test.ts:110]: Added `[os]` to `common/config/rush/build-cache.json` in 8866cb5ec3. Validated the configured pattern with the existing `CacheEntryId.parsePattern` implementation and simulated `process.platform` values: identical project/phase/input hashes produce distinct `linux`, `darwin`, and `win32` entry names. The local 931-test suite actually executed with the new pattern rather than restoring cached results.

Scope tradeoff: this uses the supported repository-wide pattern, so it partitions all phases by OS and prevents reuse of entries under the previous pattern, not just this project’s test entries. It does not change shipped Rush cache defaults. This scope is also documented in the PR description. Fresh Linux/Windows CI is pending for this commit; I have not run Windows locally.

File: common/changes/@microsoft/rush/fix-pnpm-posix-credential-environment_2026-09-18.json (hunk 1 above)
Comment (mikedelgaudio) [reply common/changes/@microsoft/rush/fix-pnpm-posix-credential-environment_2026-09-18.json]: Updated the existing change entry in 8866cb5ec3 to put `provideNpmrcCredentialsViaEnvironment` in backticks, using the suggested text. `rush change --verify --target-branch origin/main` passes.

File: libraries/rush-lib/src/utilities/test/Utilities.test.ts (hunk 2 above)
Comment (mikedelgaudio) [reply libraries/rush-lib/src/utilities/test/Utilities.test.ts]: Updated in 8866cb5ec3: the fixture now lives at the deterministic project-local `temp/utilities-credential-environment-test` path. Confirmed `../../..` resolves to the Rush-library package root from the compiled test layout. Setup and teardown delete only this suite-specific directory, and the `check credentials.cjs` filename retains the path-with-spaces coverage. All 931 Rush-library tests ran and passed on macOS; fixture cleanup was also verified.

**Comment by mikedelgaudio** [comment]
Validation update for 8866cb5ec3: all six CI jobs passed (Linux: Node 20/22/24/26; Windows: Node 24/26), and the CLA check passed. CI run: https://github.com/microsoft/rushstack/actions/runs/35648488938

Confirmed the first Windows Node 26 `Rush retest` step actually executed `Utilities.test.js` (12 passed, zero failures), not a cross-OS cache restoration. The Rush-library suite reported 930 successes out of 931 total, with the POSIX-only argument test skipped as intended. Local macOS validation passed all 931 tests.

All five review comments have implementation replies. The repository-wide OS cache partitioning tradeoff remains documented for review; no threads were marked resolved on behalf of the reviewer.

**Review by iclanton — APPROVED** [review]
